src/db-adapters/Mongoose/MongooseAdapter.js

```diff
@@ -603,30 +604,40 @@ export default class MongooseAdapter {
     return words.join(" ");
   }
 
-  static getIdQueryType(idOrIds) {
-    const mode = typeof idOrIds === "string" ? "findOne" : "find";
-    let idQuery;
-
-    if (typeof idOrIds === "string") {
-      if (!this.idIsValid(idOrIds)) {
-        throw new APIError(404, undefined, "No matching resource found.", "Invalid ID.");
-      }
-
-      idQuery = {_id: idOrIds};
+  static getIdQueryType(idOrIds, Model) {
+    if (idOrIds == null) {
+      return Q([ "find", undefined ]);
     }
 
-    else if (Array.isArray(idOrIds)) {
-      if (!idOrIds.every(this.idIsValid)) {
-        throw new APIError(400, undefined, "Invalid ID.");
-      }
-
-      idQuery = {_id: {"$in": idOrIds}};
-    }
+    const [ mode, ids ] = Array.isArray(idOrIds) ? [ "find", idOrIds ] : [ "findOne", [ idOrIds ] ];
+    const idValidationPromises = ids.map(id => this.idIsValid(id, Model));
 
-    return [mode, idQuery];
+    return Q.all(idValidationPromises).then((idsAreValid) => {
```

Comment:
> We could maintain the semantics of idIsValid() by doing something like this

That's a pretty clever solution! I like it.

Still, though, I think I'd slightly prefer just renaming `idIsValid` so that settling the promise with richer values (i.e. the id and the error, respectively) then makes sense. I lean that way only because we have the validation error information within that function, so we might as well not throw it away (in case future callers want to do something with it).